File: neetcode150/17_math_and_geometry/0201_bitwise_and_of_numbers_range.cpp
/*
 * =====================================================================
 * LeetCode 201 : Bitwise AND of Numbers Range                       (Medium)
 * https://leetcode.com/problems/bitwise-and-of-numbers-range/
 * Category   : Math / bit manipulation
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Bitwise AND of EVERY integer in [left, right] inclusive.
 * ---------------------------------------------------------------------
 * INTUITION
 *   For a bit position to be 1 in the result, ALL numbers in the
 *   range must share that bit — i.e. the common PREFIX of left and
 *   right. Everything past the first differing bit flips in between,
 *   killing it. So: right &= right-1 repeatedly until right ≤ left,
 *   or equivalently shift both until equal, then shift back.
 * ---------------------------------------------------------------------
 * APPROACH 1 · Clear trailing bits (best)
 *   while left < right: right &= right - 1  → right is now the common
 *   prefix of the two endpoints.
 * APPROACH 2 · Shift-and-compare
 *   shift = 0; while left != right: left >>= 1, right >>= 1, shift++.
 *   Answer = left << shift.
 * ---------------------------------------------------------------------
 * DEEP DIVE · Why the common prefix is the answer
 *   left and right agree on the leading bits; at the first bit where
 *   they differ, the range contains a number with a 0 there (the
 *   "carry boundary"), and below that bit the range covers every
 *   combination — so those lower bits are all 0 in the AND. The
 *   prefix bits are constant across the range, hence survive.
 *   right & (right-1) clears exactly the lowest set bit — repeating
 *   it converges to the largest prefix ≤ right that is a prefix of
 *   left... the loop stops when right ≤ left, and with right ≥ left
 *   (invariant) it lands on the prefix itself.
 * ---------------------------------------------------------------------
 * TIME COMPLEXITY : O(1) — ≤ 31 iterations.
 * MEMORY COMPLEXITY: O(1).
 * =====================================================================
 */
#include <iostream>
#include <chrono>
#include <sys/resource.h>

class Solution_1 {
public:
    int rangeBitwiseAnd(int left, int right) {
        while (left < right) right &= right - 1;   // clear lowest set bit
        return right;                              // = common prefix
    }
};

class Solution_2 {
public:
    int rangeBitwiseAnd(int left, int right) {
        int shift = 0;
        while (left != right) {                    // find common prefix
            left >>= 1;
            right >>= 1;
            ++shift;
        }
        return left << shift;                      // shift the prefix back
    }
};

static long long nowUs() {
    return std::chrono::duration_cast<std::chrono::microseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
}
static long memKb() { struct rusage r{}; getrusage(RUSAGE_SELF, &r); return r.ru_maxrss; }

int main() {
    struct Case { int l, r, want; };
    const Case cases[] = {
        {5, 7, 4},
        {0, 0, 0},
        {1, 2147483647, 0},
        {1073741824, 2147483646, 1073741824}
    };
    const long mem0 = memKb();
    auto t0 = nowUs();
    bool allOk = true;
    for (const auto& c : cases) {
        int r1 = Solution_1().rangeBitwiseAnd(c.l, c.r);
        int r2 = Solution_2().rangeBitwiseAnd(c.l, c.r);
        bool ok = r1 == c.want && r2 == c.want;
        allOk = allOk && ok;
        std::cout << "[" << c.l << "," << c.r << "] → A1=" << r1 << " A2=" << r2
                  << " (want " << c.want << ") " << (ok ? "PASS" : "FAIL") << "\n";
    }
    auto t1 = nowUs();
    std::cout << (allOk ? "PASS : all cases" : "FAIL") << " | time: " << (t1 - t0)
              << " us | mem: " << (memKb() - mem0) << " KB\n";
    return 0;
}

File: neetcode150/17_math_and_geometry/0201_bitwise_and_of_numbers_range.py
"""
=====================================================================
LeetCode 201 : Bitwise AND of Numbers Range                       (Medium)
https://leetcode.com/problems/bitwise-and-of-numbers-range/
Category   : Math / bit manipulation
---------------------------------------------------------------------
PROBLEM
    Bitwise AND of every integer in [left, right] inclusive.
---------------------------------------------------------------------
INTUITION
    A result bit is 1 only if ALL numbers in the range have it —
    i.e. it lies in the common PREFIX of left and right. Bits past
    the first difference flip inside the range and die.
---------------------------------------------------------------------
APPROACH 1 · Clear the lowest set bit (best)
    while left < right: right &= right - 1. Converges to the common
    prefix.
APPROACH 2 · Shift both until equal, shift back
    Shift left/right rightwards until they match; the prefix shifted
    back left is the answer.
---------------------------------------------------------------------
DEEP DIVE · Why the prefix survives
    Where left and right first differ, the range spans the carry
    boundary (e.g. 5=101 to 7=111: the middle bit flips) — so that
    bit and everything below is 0 in the AND. Only bits where both
    endpoints (hence every number) agree stay 1. right&(right-1)
    removes the lowest set bit; iterating until right ≤ left lands
    exactly on the shared prefix.
---------------------------------------------------------------------
TIME COMPLEXITY : O(1) — at most 31 iterations.
MEMORY COMPLEXITY: O(1)
=====================================================================
"""


class Solution_1:
    def rangeBitwiseAnd(self, left, right):
        while left < right:
            right &= right - 1
        return right


class Solution_2:
    def rangeBitwiseAnd(self, left, right):
        shift = 0
        while left != right:
            left >>= 1
            right >>= 1
            shift += 1
        return left << shift


if __name__ == "__main__":
    import time, tracemalloc
    cases = [(5, 7, 4), (0, 0, 0), (1, 2147483647, 0), (1073741824, 2147483646, 1073741824)]
    tracemalloc.start()
    t0 = time.perf_counter()
    ok = True
    for l, r, want in cases:
        r1, r2 = Solution_1().rangeBitwiseAnd(l, r), Solution_2().rangeBitwiseAnd(l, r)
        ok &= r1 == r2 == want
        print(f"[{l},{r}] -> A1={r1} A2={r2} (want {want}) {'PASS' if r1 == want and r2 == want else 'FAIL'}")
    dt = time.perf_counter() - t0
    mem = tracemalloc.get_traced_memory()[1] / 1024
    tracemalloc.stop()
    print(f"{'PASS : all cases' if ok else 'FAIL'} | time: {dt*1e6:.0f} us | mem: {mem:.0f} KB")
